Shibboleth dynamic login handler

jchssystems opened the pull request:
This feature allows to add Shibboleth users in a dynamic way from configuration.

- Configuration is located in **mh-default-org.xml**.
- Shibboleth attributes are mapped using SpEL.
- It uses a new class DynamicLoginHander instead of ConfigurableLoginHandler.

jchernandezr committed:
refs #502450 ADD Shibboleth Dynamic Loginhandler

  refs #502450 ADD Shibboleth Dynamic Loginhandler

  refs #501517 FIX Update group memberships in Shibboleth LoginHandler

refs #501517 ADD Dynamic login handler based on spring expressions

  refs #501517 ADD Tests for dynamic login handler based on spring expressions

  refs #501517 FIX Interfaces and maven-dependency-plugin

refs #4870 FIX map email and name

refs #4943 FIX Fallback to user create when there is no user reference for existing users

  - FIXME Check for existence of non-ref user first

  FIX securitty.aai Checkstyle

  FIX userdirectory Checkstyle

  FIX all

  refs #502450 FIX Shibboleth Dynamic Loginhandler

refs #2965 FIX Example configuration for Dynamic loginhandler

refs #2965 FIX Use spring util 3.1 xsd - not tested

  - FIXME We may need to update pom.xml

refs #2965 FIX Mockdata

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/api/AttributeMapper.java:
```diff
@@ -0,0 +1,186 @@
+    List<String> expressions = attributeMap.get(mappingId);
+
+    if (expressions == null) {
+      logger.error("No mapping for \"" + mappingId
+          + "\" specified. Did you forget to configure a <util:map id=\""
+          + mappingId + "\" ...?");
+      throw new IllegalArgumentException("No mapping for \"" + mappingId
+          + "\" specified. Did you forget to configure a <util:map id=\""
+          + mappingId + "\" ...?");
```
Wouldn't the exception end up in the log anyway and make the complete error message unnecessary?

jchssystems replied:
Fix

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/api/AttributeMapper.java:
```diff
@@ -0,0 +1,186 @@
+      try {
+        exp = parser.parseExpression(expression);
+        String res = (String) exp.getValue(sourceAttributes);
+        logger.debug("Mapping " + exp.getExpressionString() + " -> " + res);
```
This kind of change applies to a number of log messages in this patch:

```suggestion
        logger.debug("Mapping {} to {}", exp.getExpressionString(), res);
```

lkiesow commented on modules/userdirectory/src/main/java/org/opencastproject/userdirectory/JpaGroupRoleProvider.java:
```diff
@@ -275,6 +303,10 @@ public void updateGroupMembershipFromRoles(String userName, String orgId, List<S
     // List of the user's groups
     List<JpaGroup> membership = UserDirectoryPersistenceUtil.findGroupsByUser(userName, orgId, emf);
     for (JpaGroup group : membership) {
+      if (prefix != null && !"".equals(prefix) && !group.getRole().startsWith(prefix)) {
```
Maybe something like:
```suggestion
      if (StringUtils.isNotBlank(prefix) && !group.getRole().startsWith(prefix)) {
```

lkiesow commented on modules/userdirectory/src/main/java/org/opencastproject/userdirectory/JpaGroupRoleProvider.java:
```diff
@@ -303,7 +335,8 @@ public void updateGroupMembershipFromRoles(String userName, String orgId, List<S
         } catch (UnauthorizedException e) {
-          logger.warn("Unable to add user {} to group {} - unauthorized", userName, group.getRole());
+            logger.warn("Unable to add user {} to group {} - unauthorized: {}", userName, group.getRole(),e.getMessage(),e);
+            e.printStackTrace();
```
```suggestion
            logger.warn("Unauthorized to add user {} to group {}", userName, group.getRole(), e);
```

lkiesow commented on modules/userdirectory/src/main/java/org/opencastproject/userdirectory/JpaUserReferenceProvider.java:
```diff
@@ -159,11 +177,14 @@ public String toString() {
-    if (user != null)
-      roles.addAll(user.getRoles());
-    return roles;
+    if (roleProvider == null) {
```
Easier to swap this (less nesting):
```suggestion
    if (roleProvider != null) {
      return roleProvider.getRolesForUser(userName);
    }
```

lkiesow commented on modules/userdirectory/src/main/java/org/opencastproject/userdirectory/api/GroupRoleProvider.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.Iterator;
+import java.util.List;
+
+public interface GroupRoleProvider extends GroupProvider, RoleProvider {
```
_Comment to myself_: Not entirely sure what the GroupRoleProvider is for. Need to investigate…

hsssystems replied:
With the interface we avoid cglib proxies.. that can cause weired problems with spring..

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/api/AttributeMapper.java:
```diff
@@ -0,0 +1,186 @@
+   */
+  public List<String> getMappedAttributes(
+      Map<String, List<String>> sourceAttributes, String mappingId) {
+    Set<String> mappedAttributes = new LinkedHashSet<String>();
```
Just wondering what the expected behavior is when an attribute appears twice?

hsssystems replied:
The First one will be taken. We could add a test...

jchssystems replied:
We have a Test for a duplicate attribute

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+ * Note that this configurable implementation aims at requiring the minimum number of Shibboleth attributes
+ * to make Opencast work with most Shibboleth-based Authentication and Authorization Infrastractures (AAI).
+ */
+//public class DynamicLoginHandler implements ShibbolethLoginHandler, RoleProvider, ManagedService, InitializingBean {
```
Should this be deleted?

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+
+  @Override
+  public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
+    // TODO Auto-generated method stub
```
Why register a managed service if you don't handle any configuration?

lkiesow commented on modules/security-aai/pom.xml:
```diff
@@ -59,10 +70,12 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-dependency-plugin</artifactId>
       </plugin>
+      -->
```
Add exceptions if necessary but don't just uncomment this
```suggestion
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-dependency-plugin</artifactId>
      </plugin>
```

jchssystems replied:
Exceptions were added.

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+   */
+  private String extractName(HttpServletRequest request) {
+    String displayName = extractDisplayName(request);
+    if (displayName != null && !"".equals(displayName)) {
```
```suggestion
    if (StringUtils.isNotBlank(displayName)) {
```

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+    String givenName = "";
+    if (givenNames.size() > 0) {
+      String givenNameValue = givenNames.get(0);
+      givenName = StringUtils.isBlank(givenNameValue) ? ""
+              : new String(givenNameValue.getBytes(StandardCharsets.ISO_8859_1),
+                      StandardCharsets.UTF_8);
```
Is the input guaranteed to be ISO_8859_1?

lkiesow replied:
And if `givenNameValue` is blank wouldn't `new String(givenNameValue.getBytes(StandardCharsets.ISO_8859_1),                      StandardCharsets.UTF_8);` still be a blank string?

hsssystems replied:
Wenn givenNameValue is blank, givenName is Blank (""), otherwise not. Same code in ConfigurableLoginHandler and for the other attributes. Should be OK.

The whole building of the displayName is only a fallback since it's also possible to configure the building of the displayName in SPeL configuration...

jchssystems replied:
It is not necesary

jchssystems replied:
It seems that ISO_8859_1 is guaranteed:
https://issues.shibboleth.net/jira/projects/SSPCPP/issues/SSPCPP-2?filter=allopenissues
but the code here is unnecesary

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+              : new String(mailValue.getBytes(StandardCharsets.ISO_8859_1),
+                      StandardCharsets.UTF_8);
+      return mail;
+    } else {
+      return null;
+    }
```
```suggestion
    }
    return null;
```

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+      }
+    }
+    roles.add(new JpaRole(organization.getAnonymousRole(), organization));
+    if (StringUtils.equals(id, bootstrapUserId)) {
+      roles.add(new JpaRole(GLOBAL_ADMIN_ROLE, organization));
+    }
```
What's the rule for assigning the global admin role here?

staubesv replied:
Hi @lkiesow, long time no see... As I originally wrote this code, I can give some comments here.

To answer your question: https://github.com/opencast/opencast/blob/develop/etc/org.opencastproject.security.aai.ConfigurableLoginHandler.cfg#L12

@jchssystems Assuming that the DynamicLoginHanlder replaces the ConfigurableLoginHandler, the bootstrap user logic is obsolete as you can configure a bootstrap user directly in the security configuration.
But the following changes seem missing to me:
- Delete ConfigurableLoginHandler.java
- Delete the configuration file of the ConfigurableLoginHandler (https://github.com/opencast/opencast/blob/develop/etc/org.opencastproject.security.aai.ConfigurableLoginHandler.cfg)

Also, please properly adjust the documentation (https://github.com/opencast/opencast/blob/develop/docs/guides/admin/docs/configuration/security.aai.md). 

- Step 1 and 2 need to be revised quite a bit
- Don't forget to mention which are the mandatory headers that need to be configured in the "new way" in the security configuration to make Opencast work at all and which ones are optional
- Please explain how bootstrapping works (even if the example in the security configuration makes this obvious to me)

@JulianKniephoff This change should be mentioned in https://docs.opencast.org/develop/admin/#upgrade/#_top (Upgrading 8.x to 9.x) as it is not obvious that adaptors which use Shibboleth need to adapt their configuration quite a bit.

JulianKniephoff replied:
@staubesv good to hear from you again, and thanks for the heads up. :smile: Will do when/if this is merged.

hsssystems replied:
> What's the rule for assigning the global admin role here?

Backward compatibility. However it is also possible to assign this ROLE in the SpEL configuration, so we will remove `bootstrapUserId`

hsssystems replied:
> * Delete ConfigurableLoginHandler.java
> 
> * Delete the configuration file of the ConfigurableLoginHandler (https://github.com/opencast/opencast/blob/develop/etc/org.opencastproject.security.aai.ConfigurableLoginHandler.cfg)

It is still possible to use the old ConfigurableLoginHandler. Our intention was a new / additional LoginHandler, not to replace the old one.

staubesv replied:
@jchssystems As far as I understand, the DynamicLoginHandler offers a superset of the functionality of the ConfigureableLoginHandler. Why should Opencast include both the ConfigurableLoginHandler and the DynamicLoginHandler? That is just more code with no added value...

hsssystems replied:
@staubesv we did not want to break existing configurations for backward compatibility, not every adopter needs this functionality. My proposal: we continue with both handlers, first writing a dedicated documentation for the DynamicLoginHandler, fixing all issues and finally make a new PR to remove the ConfigurableLoginHandler

staubesv replied:
@jchssystems No strong opinion, but you could expect adopters to change the configuration. Note, however, that "fixing all issues" should be done before merging ;-)

jchssystems replied:
bootstrapUserId was removed.
A new independent documentation will be made when all the code be fixed.

jchssystems replied:
In my opinion you should not force to use one or another login handler (or a third in the future).

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+  private JpaOrganization fromOrganization(Organization org) {
+    if (org instanceof JpaOrganization) {
+      return (JpaOrganization) org;
+    } else {
+      return new JpaOrganization(org.getId(), org.getName(), org.getServers(), org.getAdminRole(),
+           org.getAnonymousRole(), org.getProperties());
+    }
```
```suggestion
    }
    return new JpaOrganization(org.getId(), org.getName(), org.getServers(), org.getAdminRole(),
         org.getAnonymousRole(), org.getProperties());
```

lkiesow commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+    this.attributeMapper = attributeMapper;
+  }
+
+  @Override
+  public List<Role> getRolesForGroup(String groupName) {
+    // TODO Auto-generated method stub
+    return null;
+  }
+
+  @Override
+  public void updateGroupMembershipFromRoles(String userName, String orgId, List<String> roleList) {
+    // TODO Auto-generated method stub
+  }
+
+  @Override
+  public void addGroup(JpaGroup group) throws UnauthorizedException {
+    // TODO Auto-generated method stub
+  }
+
+  @Override
+  public Iterator<Group> getGroups() {
+    // TODO Auto-generated method stub
+    return null;
+  }
+
+  @Override
+  public void createGroup(String name, String description, String roles, String users)
+         throws IllegalArgumentException, UnauthorizedException, ConflictException {
+    // TODO Auto-generated method stub
+  }
+
+  @Override
+  public void updateGroup(String groupId, String name, String description, String roles, String users)
+         throws NotFoundException, UnauthorizedException {
+    // TODO Auto-generated method stub
+  }
```
There are a lot of to-dos

jchssystems replied:
Fix

lkiesow reviewed (changes requested):
I haven't actually compiled or even tested this but just skimmed through the code. I didn't find any critical issues, just a bunch of minor comments so far…

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+ */
+package org.opencastproject.security.aai;
+
+import static org.opencastproject.security.api.SecurityConstants.GLOBAL_ADMIN_ROLE;
```
Obsolete if removing bootstrap user

jchssystems replied:
Removed

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+public class DynamicLoginHandler implements ShibbolethLoginHandler, GroupRoleProvider, ManagedService, InitializingBean {
+
+  /** Default value of the configuration property CFG_ROLE_FEDERATION_KEY */
+  private static final String CFG_ROLE_FEDERATION_DEFAULT = "ROLE_AAI_USER";
```
Leftover from old code

jchssystems replied:
CFG_ROLE_FEDERATION_DEFAULT was removed

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+  private AttributeMapper attributeMapper = null;
+
+  /** The ID of the bootstrap user if configured */
+  private String bootstrapUserId = null;
```
Obsolete when removing boot strap user

jchssystems replied:
bootstrapUserId was removed.

jchssystems replied:
bootstrapUserId was also removed

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+  private String bootstrapUserId = null;
+
+  /** Role assigned to all Shibboleth authenticated users */
+  private String roleFederationMember = CFG_ROLE_FEDERATION_DEFAULT;
```
Leftover from old version

jchssystems replied:
CFG_ROLE_FEDERATION_DEFAULT was also removed

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+  private String roleFederationMember = CFG_ROLE_FEDERATION_DEFAULT;
+
+  /*
+   * It is the bundle matterhorn-kernel what will need to instantiate the ConfigurableLoginHandler
```
DynamicLoginHandler

hsssystems commented on modules/userdirectory/src/main/java/org/opencastproject/userdirectory/JpaGroupRoleProvider.java:
```diff
@@ -81,7 +82,7 @@
   service = { RoleProvider.class, JpaGroupRoleProvider.class }
 )
-public class JpaGroupRoleProvider extends AbstractIndexProducer implements RoleProvider, GroupProvider {
+public class JpaGroupRoleProvider extends AbstractIndexProducer implements RoleProvider, GroupProvider, GroupRoleProvider {
```
Isn't it sufficient to implement only GroupRoleProvider?

staubesv commented on modules/security-aai/src/main/java/org/opencastproject/security/aai/DynamicLoginHandler.java:
```diff
@@ -0,0 +1,432 @@
+      }
+    }
+    roles.add(new JpaRole(organization.getAnonymousRole(), organization));
+    if (StringUtils.equals(id, bootstrapUserId)) {
```
 @jchssystems bootstrapUserId is loaded from the configuration file in the original ConfigurableLoginHandler. As you don't it in DynamicLoginHandler, bootstrapUserId will always be null. But since you can configure a bootstrap user in the security configuration, it is safe to just omit those lines here.

jchssystems commented:
Please, review tthe code again. Some changes were made.
I'm still not sure if both logins should exist or if the new one will replace the old one.
And about the documentation, should I have a totally independent page for the new login handler? when the code is ok, fix the documentation.

Commit:
refs #5049 FIX Do not use sn or givenName since we have only displayName in OC

gregorydlogan commented:
@staubesv any chance you can take another pass through this?  I'm unfamiliar with any other major adopter using Shibboleth.

hsssystems commented:
> @staubesv any chance you can take another pass through this? I'm unfamiliar with any other major adopter using Shibboleth.

Yesterday we made another integration test on a OC9 test installation with this branch and it worked. However, we found some issues:

- we need to adapt the the documentation (example configuration) and the unit test
- Updating the displayName or group memberships (only removing users from groups) needs several logins. Do we need some index / cache invalidation?
- The GroupRoleProvider interface needs refactoring

otti-ssystems committed:
refs #2965 REFACTOR structure of interfaces

otti-ssystems committed:
refs #2965 REFACTOR deleting unnecessary variables

otti-ssystems committed:
refs #2965 REFACTOR Interfaces and config for the Dynamic Login Handler

otti-ssystems commented:
Hi.

I adjusted the structure of the interfaces.
DynamicLoginHandler is not extending GroupRoleProvider anymore. 
Instead we are using now AAIRoleProvider which contains getRoles, so 
i haven't had to touch RoleProvider.

I also ajusted the Tests, because the mapping of givenName and sn is not 
needed in this context.

Travis is not able to build because of a weird Exception.

https://travis-ci.com/github/opencast/opencast/jobs/366125028

jchernandezr committed:
refs #2965 FIX Set html Entities in java documentation

jchssystems commented:
@lkiesow  @gregorydlogan
I have fixed the problem that was there some time ago and it compiles well.  The code was also refactored. Could you please check it again? So we can close the issue as soon as possible.
Thanks in advance.

otti-ssystems commented:
Hi Guys.
Sorry but i dont get it.

Is there still something to be done or is this ready to merge?

gregorydlogan commented:
Hi @jchssystems @otti-ssystems we don't get notifications when things change, so unless someone manually checks we don't know that you'd pushed changes.  If you don't see something happen within a week or so please ping the people who have been active in the ticket, or pipe up on the dev list.

I have no way to test this unfortunately, so I"m somewhat hamstrung in terms of getting it merged.  I'm assuming you're writing this for a client, can they verify that it's working in prod?

hsssystems commented:
Hi Greg,

the handler is in production on several clients! However, we will deploy again an Opencast 9 with the handler and test it another time systematically (we did already twice). Is that OK for you? 

Our QA guys will test it, not the programmers, so it should be safe :-)

gregorydlogan commented:
Ok, sounds fair. Let's merge this then.
